src/Microsoft.IdentityModel.Abstractions/EventDetails.cs

```diff
@@ -0,0 +1,197 @@
+        {
+            BoolPropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.Bool;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Sets a property on the event details.
+        /// </summary>
+        /// <param name="key">Property key.</param>
+        /// <param name="value">Property value.</param>
+        /// <param name="dataClassification">Optional parameter which indicates the classification of the data.</param>
+        public virtual void SetProperty(
+            string key,
+            DateTime value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            DateTimePropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.DateTime;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Sets a property on the event details.
+        /// </summary>
+        /// <param name="key">Property key.</param>
+        /// <param name="value">Property value.</param>
+        /// <param name="dataClassification">Optional parameter which indicates the classification of the data.</param>
+        public virtual void SetProperty(
+            string key,
+            double value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            DoublePropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.Double;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Sets a property on the event details.
+        /// </summary>
+        /// <param name="key">Property key.</param>
+        /// <param name="value">Property value.</param>
+        /// <param name="dataClassification">Optional parameter which indicates the classification of the data.</param>
+        public virtual void SetProperty(
+            string key,
+            Guid value,
+            DataClassification dataClassification = DataClassification.SystemMetadata)
+        {
+            GuidPropertyValues[key] = value;
+            PropertiesKeys[key] = DictionaryTypeEnum.Guid;
+            AddDataClassificationIfNecessary(key, dataClassification);
+        }
+
+        /// <summary>
+        /// Updates the state of the property data classification.
+        /// </summary>
+        /// <param name="key">Key of the property.</param>
+        /// <param name="dataClassification">DataClassification for the property.</param>
+        internal void AddDataClassificationIfNecessary(string key, DataClassification dataClassification)
+        {
+            if (dataClassification != DataClassification.SystemMetadata)
+                DataClassificationMapping[key] = dataClassification;
+            else
+                DataClassificationMapping.Remove(key);
```

Comment:
I believe this is not thread safe. You can either use ConcurrentDictionary operations or a lock (since this is pretty much a corner case?)